### PR TITLE
#13: Join support.

### LINKED ISSUE #13: Joins
## Situation

Join and sub-select rendering are available in the base library, but the dialect does not report the capabilities and integration tests are missing.

Sub-selects are unfortunately not supported by the core database, so the adapter can only handle joins.

## Acceptance Criteria
1. Integration tests prove joins are supported

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #13.

### CODE REVIEW

File: doc/changes/changes_0.2.0.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Exasol Row Level Security (Lua) 0.2.0, released 2022-07-27
```
Comment (kaklakariada) [inline doc/changes/changes_0.2.0.md:1]:
```suggestion
# Exasol Row Level Security (Lua) 0.2.0, released 2022-07-28
```
or -29?

File: src/test/java/com/exasol/SelectIT.java [hunk 2]
```diff
@@ -62,8 +72,76 @@ void testSelectWithOrderByExpressionAndLimitWithOffset() {
...
+    @Test
+    void testInnerJoin() {
+        final Schema sourceSchema = createJoinSchema("INNER_JOIN_SCHEMA");
+        final VirtualSchema virtualSchema = createVirtualSchema(sourceSchema);
+        final User user = createUserWithVirtualSchemaAccess("INNER_JOIN_USER", virtualSchema);
+        final String virtualSchemaName = getVirtualSchemaName(sourceSchema.getName());
+        assertJoinQuery("SELECT LJC, RJC FROM " + virtualSchemaName + ".T_LEFT"
+                + " INNER JOIN " + virtualSchemaName + ".T_RIGHT ON LJC = RJC ORDER BY LJC, RJC", user,
+                table().row("L+R", "L+R"), "SELECT.*FROM.*INNER JOIN.*");
+    }
```
Comment (kaklakariada) [inline src/test/java/com/exasol/SelectIT.java:89]: 👍 Tests are easy to understand!

File: src/main/lua/exasolvs/adapter_capabilities.lua [hunk 3]
```diff
@@ -4,7 +4,8 @@ return {
     -- Main capabilities
     "SELECTLIST_PROJECTION", "SELECTLIST_EXPRESSIONS", "AGGREGATE_GROUP_BY_COLUMN", "AGGREGATE_GROUP_BY_TUPLE",
     "AGGREGATE_GROUP_BY_EXPRESSION", "AGGREGATE_SINGLE_GROUP", "LIMIT", "LIMIT_WITH_OFFSET", "ORDER_BY_COLUMN",
-    "ORDER_BY_EXPRESSION",
+    "ORDER_BY_EXPRESSION", "JOIN", "JOIN_CONDITION_EQUI", "JOIN_TYPE_INNER",  "JOIN_TYPE_FULL_OUTER",
```
Comment (kaklakariada) [inline src/main/lua/exasolvs/adapter_capabilities.lua:7]: What does `JOIN_CONDITION_EQUI` mean? Do we need an integration test for that?

File: src/main/lua/exasolvs/QueryRewriter.lua [hunk 4]
```diff
@@ -43,8 +43,32 @@ local function rewrite(query)
...
+    if(type(element) == "table") then
+        for key, value in pairs(element) do
+            if(type(value) == "table") then
+                extended_element[key] = extend_query_element_with_source_schema(value, source_schema_id)
+            else
+                extended_element[key] = value
+            end
+        end
+        if(element.type ~= nil and element.type == "table" and element.schema == nil) then
+            log.debug("Extending: " .. element.name)
```
Comment (kaklakariada) [inline src/main/lua/exasolvs/QueryRewriter.lua:57]:
```suggestion
            log.debug("Extending: %s", element.name)
```
More efficient?

File: src/main/lua/exasolvs/QueryRewriter.lua [hunk 5]
```diff
@@ -43,8 +43,32 @@ local function rewrite(query)
     expand_select_list(query)
 end
 
+local function extend_query_element_with_source_schema(element, source_schema_id)
```
Comment (kaklakariada) [inline src/main/lua/exasolvs/QueryRewriter.lua:46]: Do we need unit tests for this function or are the integration tests enough?

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/lua/exasolvs/adapter_capabilities.lua (hunk 3 above)
Comment (redcatbear) [reply src/main/lua/exasolvs/adapter_capabilities.lua:7]: That's a fancy way of saying you support a "=" in the join condition.

File: doc/changes/changes_0.2.0.md (hunk 1 above)
Comment (redcatbear) [reply doc/changes/changes_0.2.0.md:1]: Updated.

File: src/main/lua/exasolvs/adapter_capabilities.lua (hunk 3 above)
Comment (redcatbear) [reply src/main/lua/exasolvs/adapter_capabilities.lua:7]: You reminded me that I forgot the complex join condition capability. Added that and a matching integration test.

File: src/main/lua/exasolvs/QueryRewriter.lua (hunk 4 above)
Comment (redcatbear) [reply src/main/lua/exasolvs/QueryRewriter.lua:57]: Fixed.

File: src/main/lua/exasolvs/QueryRewriter.lua (hunk 5 above)
Comment (redcatbear) [reply src/main/lua/exasolvs/QueryRewriter.lua:46]: A unit test helps locate errors better, so I think keeping both is worth it. Also it showcases what this workaround is for.

File: src/main/lua/exasolvs/adapter_capabilities.lua (hunk 3 above)
Comment (kaklakariada) [reply src/main/lua/exasolvs/adapter_capabilities.lua]: :-) 👍

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**